tests/integrations/raftstore/test_unsafe_recovery.rs

```diff
@@ -0,0 +1,544 @@
+    // forbid reads in force leader state
+    let get = new_get_cmd(b"k1");
+    must_get_error_recovery_in_progress(&mut cluster, &region, get);
+    // forbid read index in force leader state
+    let read_index = new_read_index_cmd();
+    must_get_error_recovery_in_progress(&mut cluster, &region, read_index);
+    cluster.exit_force_leader(region.get_id(), 1);
+
+    // quorum is formed, can propose command successfully now
+    cluster.must_put(b"k4", b"v4");
+    assert_eq!(cluster.must_get(b"k2"), None);
+    assert_eq!(cluster.must_get(b"k3"), None);
+    assert_eq!(cluster.must_get(b"k4"), Some(b"v4".to_vec()));
+}
+
+// Test the case that three of five nodes fail and force leader on one of the
+// rest nodes.
+#[test]
+fn test_force_leader_five_nodes() {
+    let mut cluster = new_node_cluster(0, 5);
+    cluster.pd_client.disable_default_operator();
+
+    cluster.run();
+    cluster.must_put(b"k1", b"v1");
+
+    let region = cluster.get_region(b"k1");
+    cluster.must_split(&region, b"k9");
+    let mut region = cluster.get_region(b"k2");
+    let peer_on_store5 = find_peer(&region, 5).unwrap();
+    cluster.must_transfer_leader(region.get_id(), peer_on_store5.clone());
+
+    cluster.stop_node(3);
+    cluster.stop_node(4);
+    cluster.stop_node(5);
+
+    let put = new_put_cmd(b"k2", b"v2");
+    let req = new_request(region.get_id(), region.take_region_epoch(), vec![put], true);
+    // quorum is lost, can't propose command successfully.
+    assert!(
+        cluster
+            .call_command_on_leader(req, Duration::from_millis(10))
+            .is_err()
+    );
+
+    cluster.enter_force_leader(region.get_id(), 1, HashSet::from_iter(vec![3, 4, 5]));
+    // remove the peers on failed nodes
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 3).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 4).unwrap().clone());
+    cluster
+        .pd_client
+        .must_remove_peer(region.get_id(), find_peer(&region, 5).unwrap().clone());
+    // forbid writes in force leader state
+    let put = new_put_cmd(b"k3", b"v3");
+    must_get_error_recovery_in_progress(&mut cluster, &region, put);
+    // forbid reads in force leader state
+    let get = new_get_cmd(b"k1");
+    must_get_error_recovery_in_progress(&mut cluster, &region, get);
+    // forbid read index in force leader state
+    let read_index = new_read_index_cmd();
+    must_get_error_recovery_in_progress(&mut cluster, &region, read_index);
+
+    cluster.exit_force_leader(region.get_id(), 1);
+
+    // quorum is formed, can propose command successfully now
+    cluster.must_put(b"k4", b"v4");
+    assert_eq!(cluster.must_get(b"k2"), None);
+    assert_eq!(cluster.must_get(b"k3"), None);
+    assert_eq!(cluster.must_get(b"k4"), Some(b"v4".to_vec()));
+}
+
+// Test the case that three of five nodes fail and force leader on the rest node
+// which is a learner.
```

Comment:
How about joint state?

Reply:
The case for joint consesus would be elaborated in this PR https://github.com/tikv/tikv/pull/12022